host/traceh.c

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/trace.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#if defined(__linux__)
+#include <sys/time.h>
+#endif
+#include <time.h>
+#include "enclave.h"
+#include "hostthread.h"
+
+#define LOGGING_FORMAT_STRING "%02d:%02d:%02d:%06ld TID(0x%lx) (%s)[%-5s]%s"
+static char* log_level_strings[OE_LOG_LEVEL_MAX] = {"NONE",
+                                                    "FATAL",
+                                                    "ERROR",
+                                                    "WARN",
+                                                    "INFO"};
+static oe_mutex g_log_lock = OE_H_MUTEX_INITIALIZER;
+static const char* g_log_file_name = NULL;
+static log_level_t g_log_level = OE_LOG_LEVEL_ERROR;
+static bool g_initialized = false;
+
+static log_level_t env2log_level(void)
+{
+    log_level_t level = OE_LOG_LEVEL_ERROR;
+    const char* level_str = getenv("OE_LOG_LEVEL");
+
+    if (level_str == NULL)
+    {
+        goto done;
+    }
+    else if (strcmp(level_str, "INFO") == 0)
+    {
+        level = OE_LOG_LEVEL_INFO;
+    }
+    else if (strcmp(level_str, "WARNING") == 0)
+    {
+        level = OE_LOG_LEVEL_WARNING;
+    }
+    else if (strcmp(level_str, "ERROR") == 0)
+    {
+        level = OE_LOG_LEVEL_ERROR;
+    }
+    else if (strcmp(level_str, "FATAL") == 0)
+    {
+        level = OE_LOG_LEVEL_FATAL;
+    }
+    else if (strcmp(level_str, "NONE") == 0)
+    {
+        level = OE_LOG_LEVEL_NONE;
+    }
+done:
+    return level;
+}
+
+static void initialize_log_config()
+{
+    if (!g_initialized)
+    {
+        // inititalize if not already
+        g_log_level = env2log_level();
+        g_log_file_name = getenv("OE_LOG_DEVICE");
+        g_initialized = true;
+    }
+}
+
+static void write_header_info_to_stream(FILE* stream)
+{
+    time_t t = time(NULL);
+    struct tm* lt = localtime(&t);
+
+    fprintf(
+        stream, "================= New logging session =================\n");
+    fprintf(stream, "%s", asctime(lt));
+    fprintf(
+        stream,
+        "https://github.com/Microsoft/openenclave branch:%s\n",
+        OE_REPO_BRANCH_NAME);
+    fprintf(stream, "Last commit:%s\n\n", OE_REPO_LAST_COMMIT);
+}
+
+static void write_message_to_stream(
+    FILE* stream,
+    bool is_enclave,
+    oe_log_args_t* args)
+{
+#if defined(__linux__)
+    struct timeval time_now;
+    gettimeofday(&time_now, NULL);
+    struct tm* t = gmtime(&time_now.tv_sec);
+#else
+    time_t lt = time(NULL);
+    struct tm* t = localtime(&lt);
+#endif
+
+    oe_thread thread_id = oe_thread_self();
+
+    fprintf(
```

Comment:
This is not thread safe. If multiple threads call this function, the output will be jumbled.